### PR TITLE
Pool Thread Cache support memory leak check.

### PR DESCRIPTION
**Pull request by horizonzy**
Motivation:
Fixes-#12749

1. If the PoolThreadCache is already freed, didn't add the ByteBuf to the cache, free it to PoolChunk directly.
2. If the PoolThreadCache leak, print warns log.

### CODE REVIEW

File: buffer/src/main/java/io/netty/buffer/PoolThreadCache.java [hunk 1]
```diff
@@ -232,6 +235,20 @@ void free(boolean finalizer) {
...
+            checkCacheIsLeak(normalDirectCaches, "NormalDirectCaches");
+            checkCacheIsLeak(smallSubPageHeapCaches, "SmallSubPageHeapCaches");
+            checkCacheIsLeak(normalHeapCaches, "NormalHeapCaches");
+        }
+    }
+
+    private void checkCacheIsLeak(MemoryRegionCache<?>[] caches, String type) {
+        for (MemoryRegionCache<?> cache : caches) {
+            if (cache.queue.size() > 0) {
+                logger.warn("[] memory may leak.", type);
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PoolThreadCache.java:250]:
```suggestion
                logger.debug("{} memory may leak.", type);
                return;
```

File: buffer/src/main/java/io/netty/buffer/PoolThreadCache.java [hunk 2]
```diff
@@ -232,6 +235,20 @@ void free(boolean finalizer) {
...
+        } else {
+            //see https://github.com/netty/netty/issues/12749
+            checkCacheIsLeak(smallSubPageDirectCaches, "SmallSubPageDirectCaches");
+            checkCacheIsLeak(normalDirectCaches, "NormalDirectCaches");
+            checkCacheIsLeak(smallSubPageHeapCaches, "SmallSubPageHeapCaches");
+            checkCacheIsLeak(normalHeapCaches, "NormalHeapCaches");
+        }
+    }
+
+    private void checkCacheIsLeak(MemoryRegionCache<?>[] caches, String type) {
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PoolThreadCache.java:247]:
```suggestion
            // See https://github.com/netty/netty/issues/12749
            checkCacheMayLeak(smallSubPageDirectCaches, "SmallSubPageDirectCaches");
            checkCacheMayLeak(normalDirectCaches, "NormalDirectCaches");
            checkCacheMayLeak(smallSubPageHeapCaches, "SmallSubPageHeapCaches");
            checkCacheMayLeak(normalHeapCaches, "NormalHeapCaches");
        }
    }

    private static void checkCacheMayLeak(MemoryRegionCache<?>[] caches, String type) {
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: buffer/src/main/java/io/netty/buffer/PoolThreadCache.java (hunk 1 above)
Comment (horizonzy) [reply buffer/src/main/java/io/netty/buffer/PoolThreadCache.java]: it's a low-level mistake. 😆

File: buffer/src/main/java/io/netty/buffer/PoolThreadCache.java [hunk 3]
```diff
@@ -179,6 +179,9 @@ private boolean allocate(MemoryRegionCache<?> cache, PooledByteBuf buf, int reqC
     @SuppressWarnings({ "unchecked", "rawtypes" })
     boolean add(PoolArena<?> area, PoolChunk chunk, ByteBuffer nioBuffer,
                 long handle, int normCapacity, SizeClass sizeClass) {
+        if (freed.get()) {
+            return false;
+        }
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PoolThreadCache.java:184]: should we move this closer to when we actually add it to the data-structure ?

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: buffer/src/main/java/io/netty/buffer/PoolThreadCache.java (hunk 3 above)
Comment (horizonzy) [reply buffer/src/main/java/io/netty/buffer/PoolThreadCache.java:184]: Fine, that's will be better.

Comment (horizonzy) [reply buffer/src/main/java/io/netty/buffer/PoolThreadCache.java:184]: If we check it in MemoryRegionCache.add, the class `MemoryRegionCache` is static, we should change it to a non-static class. I'm not sure if is it deserved.

Comment (normanmaurer) [reply buffer/src/main/java/io/netty/buffer/PoolThreadCache.java:184]: Just put the check in front of `return cache.add(chunk, nioBuffer, handle, normCapacity);`

Comment (horizonzy) [reply buffer/src/main/java/io/netty/buffer/PoolThreadCache.java]: done

**Review by chrisvest — APPROVED** [review]

**Comment by normanmaurer** [comment]
@horizonzy Thanks a lot!

**Pull request merged**